# generated by datamodel-codegen:
#   filename:  brapi_generated.json
#   timestamp: 2025-10-23T23:28:51+00:00

from __future__ import annotations

from enum import Enum
from typing import List, Optional, Union

from pydantic import AnyUrl, BaseModel, Field, RootModel


class AdditionalInfo(BaseModel):
    additionalProperties: Optional[str] = None


class AlleleMatrix(BaseModel):
    callSetDbIds: List[str]
    expandHomozygotes: Optional[bool] = None
    sepPhased: Optional[str] = None
    sepUnphased: Optional[str] = None
    unknownString: Optional[str] = None
    variantDbIds: Optional[List[str]] = None
    variantSetDbIds: List[str]


class AlleleMatrixSearchRequest(BaseModel):
    expandHomozygotes: Optional[bool] = None
    preview: Optional[bool] = None
    sepPhased: Optional[str] = None
    sepUnphased: Optional[str] = None
    unknownString: Optional[str] = None


class Analysis(BaseModel):
    analysisDbId: str
    analysisName: Optional[str] = None
    created: Optional[str] = None
    description: Optional[str] = None
    software: Optional[List[str]] = None
    type: Optional[str] = None
    updated: Optional[str] = None
    variantSetDbId: Optional[str] = None
    variantSetName: Optional[str] = None


class DataFormat(Enum):
    DartSeq = 'DartSeq'
    VCF = 'VCF'
    Hapmap = 'Hapmap'
    tabular = 'tabular'
    JSON = 'JSON'


class FileFormat(Enum):
    text_csv = 'text/csv'
    text_tsv = 'text/tsv'
    application_excel = 'application/excel'
    application_zip = 'application/zip'
    application_json = 'application/json'


class AvailableFormat(BaseModel):
    dataFormat: Optional[DataFormat] = Field(
        None,
        description='dataFormat defines the structure of the data within a file (ie DartSeq, VCF, Hapmap, tabular, etc)',
    )
    expandHomozygotes: Optional[bool] = None
    fileFormat: Optional[FileFormat] = Field(
        None,
        description='fileFormat defines the MIME type of the file (ie text/csv, application/excel, application/zip). This should also be reflected in the Accept and ContentType HTTP headers for every relevant request and response.',
    )
    fileURL: Optional[str] = None
    sepPhased: Optional[str] = None
    sepUnphased: Optional[str] = None
    unknownString: Optional[str] = None
    variantSetDbId: Optional[str] = None
    variantSetName: Optional[str] = None


class BreedingMethod(BaseModel):
    abbreviation: str
    breedingMethodDbId: str
    breedingMethodName: str
    description: str


class Call(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    callSetDbId: str
    callSetName: Optional[str] = None
    genotypeValue: Optional[str] = None
    phaseSet: Optional[str] = None
    variantDbId: str
    variantSetDbId: str
    variantSetName: Optional[str] = None


class CallSearchRequest(BaseModel):
    expandHomozygotes: Optional[bool] = None
    sepPhased: Optional[str] = None
    sepUnphased: Optional[str] = None
    unknownString: Optional[str] = None


class CallSet(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    callSetDbId: str
    callSetName: Optional[str] = None
    created: Optional[str] = None
    sampleDbId: Optional[str] = None
    sampleName: Optional[str] = None
    samplePUI: Optional[str] = None
    studyDbId: Optional[str] = None
    studyName: Optional[str] = None
    studyPUI: Optional[str] = None
    updated: Optional[str] = None
    variantSetDbIds: Optional[List[str]] = None


class CallSetSearchRequest(BaseModel):
    pass


class Contact(BaseModel):
    contactDbId: str
    email: Optional[str] = None
    instituteName: Optional[str] = None
    name: Optional[str] = None
    orcid: Optional[str] = None
    type: Optional[str] = None


class ContentMixture(BaseModel):
    crossDbId: Optional[str] = None
    crossName: Optional[str] = None
    germplasmDbId: Optional[str] = None
    germplasmName: Optional[str] = None
    germplasmPUI: Optional[str] = None
    mixturePercentage: Optional[int] = None
    seedLotDbId: Optional[str] = None
    seedLotName: Optional[str] = None


class ContentTypes(Enum):
    application_json = 'application/json'
    text_csv = 'text/csv'
    text_tsv = 'text/tsv'
    application_flapjack = 'application/flapjack'


class Context(RootModel[List[AnyUrl]]):
    root: List[AnyUrl] = Field(
        ...,
        description='The JSON-LD Context is used to provide JSON-LD definitions to each field in a JSON object. By providing an array of context file urls, a BrAPI response object becomes JSON-LD compatible.  \n\nFor more information, see https://w3c.github.io/json-ld-syntax/#the-context',
        examples=[['https://brapi.org/jsonld/context/metadata.jsonld']],
        title='context',
    )


class CrossAttribute(BaseModel):
    crossAttributeName: Optional[str] = None
    crossAttributeValue: Optional[str] = None
    crossDbId: Optional[str] = None
    crossName: Optional[str] = None


class CrossType(Enum):
    BIPARENTAL = 'BIPARENTAL'
    SELF = 'SELF'
    OPEN_POLLINATED = 'OPEN_POLLINATED'
    BULK = 'BULK'
    BULK_SELFED = 'BULK_SELFED'
    BULK_OPEN_POLLINATED = 'BULK_OPEN_POLLINATED'
    DOUBLE_HAPLOID = 'DOUBLE_HAPLOID'


class CrossingProject(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    commonCropName: Optional[str] = None
    crossingProjectDescription: Optional[str] = None
    crossingProjectName: str
    programDbId: Optional[str] = None
    programName: Optional[str] = None


class CrossingProjectNewRequest(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    commonCropName: Optional[str] = None
    crossingProjectDbId: str
    crossingProjectDescription: Optional[str] = None
    crossingProjectName: str
    programDbId: Optional[str] = None
    programName: Optional[str] = None


class DataLink(BaseModel):
    dataFormat: Optional[str] = None
    description: Optional[str] = None
    fileFormat: Optional[str] = None
    name: Optional[str] = None
    provenance: Optional[str] = None
    scientificType: Optional[str] = None
    url: Optional[str] = None
    version: Optional[str] = None


class DataType(Enum):
    string = 'string'
    integer = 'integer'
    float = 'float'
    boolean = 'boolean'


class DataMatrix(BaseModel):
    alleleMatrix: Optional[AlleleMatrix] = None
    dataMatrix: Optional[List[List[str]]] = None
    dataMatrixAbbreviation: Optional[str] = None
    dataMatrixName: Optional[str] = None
    dataType: Optional[DataType] = Field(
        None,
        description='The type of field represented in this data matrix. This is intended to help parse the data out of JSON.',
    )


class DatasetAuthorships(BaseModel):
    datasetPUI: Optional[str] = None
    license: Optional[str] = None
    publicReleaseDate: Optional[str] = None
    submissionDate: Optional[str] = None
    trialDbId: Optional[str] = None
    trialName: Optional[str] = None
    trialPUI: Optional[str] = None


class Type(Enum):
    OBO = 'OBO'
    RDF = 'RDF'
    WEBPAGE = 'WEBPAGE'


class DocumentationLink(BaseModel):
    URL: Optional[str] = None
    ontologyReferenceDbId: Optional[str] = None
    type: Optional[Type] = None


class Donor(BaseModel):
    donorAccessionNumber: Optional[str] = None
    donorInstituteCode: Optional[str] = None
    germplasmDbId: Optional[str] = None
    germplasmName: Optional[str] = None
    germplasmPUI: Optional[str] = None


class EnvironmentParameters(BaseModel):
    description: str
    environmentParametersDbId: str
    parameterName: str
    parameterPUI: Optional[str] = None
    studyDbId: Optional[str] = None
    studyName: Optional[str] = None
    studyPUI: Optional[str] = None
    unit: Optional[str] = None
    unitPUI: Optional[str] = None
    value: Optional[str] = None
    valuePUI: Optional[str] = None


class Event(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    eventDbId: str
    eventDescription: Optional[str] = None
    eventType: str
    eventTypeDbId: Optional[str] = None
    observationUnitDbIds: Optional[List[str]] = None
    studyDbId: Optional[str] = None
    studyName: Optional[str] = None
    studyPUI: Optional[str] = None


class EventDateRange(BaseModel):
    discreteDates: Optional[List[str]] = None
    endDate: Optional[str] = None
    eventDbId: Optional[str] = None
    startDate: Optional[str] = None


class EventParameter(BaseModel):
    code: Optional[str] = None
    description: Optional[str] = None
    eventDbId: Optional[str] = None
    name: Optional[str] = None
    units: Optional[str] = None
    value: Optional[str] = None
    valueDescription: Optional[str] = None
    valuesByDate: Optional[List[str]] = None


class ExperimentalDesign(BaseModel):
    PUI: Optional[str] = None
    description: Optional[str] = None
    studyDbId: Optional[str] = None
    studyName: Optional[str] = None
    studyPUI: Optional[str] = None


class ExternalReference(BaseModel):
    referenceId: Optional[str] = None
    referenceSource: Optional[str] = None


class GenomeMap(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    comments: Optional[str] = None
    commonCropName: str
    documentationURL: Optional[str] = None
    linkageGroupCount: Optional[int] = None
    mapDbId: str
    mapName: Optional[str] = None
    mapPUI: Optional[str] = None
    markerCount: Optional[int] = None
    publishedDate: Optional[str] = None
    scientificName: Optional[str] = None
    type: str
    unit: Optional[str] = None


class GenotypeMetadata(BaseModel):
    call: Optional[Call] = None
    dataType: Optional[DataType] = Field(
        None,
        description='The type of field represented in this Genotype Field. This is intended to help parse the data out of JSON.',
    )
    fieldAbbreviation: Optional[str] = None
    fieldName: Optional[str] = None
    fieldValue: Optional[str] = None


class GeoJSONGeometry1(BaseModel):
    coordinates: Optional[List[float]] = None
    type: str


class GeoJSONGeometry2(BaseModel):
    coordinates: Optional[List[List[List[float]]]] = None
    type: str


class GeoJSONGeometry(RootModel[Union[GeoJSONGeometry1, GeoJSONGeometry2]]):
    root: Union[GeoJSONGeometry1, GeoJSONGeometry2]


class BiologicalStatusOfAccessionCode(Enum):
    field_100 = '100'
    field_110 = '110'
    field_120 = '120'
    field_130 = '130'
    field_200 = '200'
    field_300 = '300'
    field_400 = '400'
    field_410 = '410'
    field_411 = '411'
    field_412 = '412'
    field_413 = '413'
    field_414 = '414'
    field_415 = '415'
    field_416 = '416'
    field_420 = '420'
    field_421 = '421'
    field_422 = '422'
    field_423 = '423'
    field_500 = '500'
    field_600 = '600'
    field_999 = '999'


class Germplasm(BaseModel):
    accessionNumber: Optional[str] = None
    acquisitionDate: Optional[str] = None
    additionalInfo: Optional[AdditionalInfo] = None
    biologicalStatusOfAccessionCode: Optional[BiologicalStatusOfAccessionCode] = Field(
        None,
        description='MCPD (v2.1) (SAMPSTAT) 19. The coding scheme proposed can be used at 3 different levels of detail: either by using the general codes such as 100, 200, 300, 400, or by using the more specific codes such as 110, 120, etc. \n\n100) Wild \n110) Natural \n120) Semi-natural/wild \n130) Semi-natural/sown \n200) Weedy \n300) Traditional cultivar/landrace \n400) Breeding/research material \n410) Breeders line \n411) Synthetic population \n412) Hybrid \n413) Founder stock/base population \n414) Inbred line (parent of hybrid cultivar) \n415) Segregating population \n416) Clonal selection \n420) Genetic stock \n421) Mutant (e.g. induced/insertion mutants, tilling populations) \n422) Cytogenetic stocks (e.g. chromosome addition/substitution, aneuploids,  amphiploids) \n423) Other genetic stocks (e.g. mapping populations) \n500) Advanced or improved cultivar (conventional breeding methods) \n600) GMO (by genetic engineering) \n999) Other (Elaborate in REMARKS field)',
    )
    biologicalStatusOfAccessionDescription: Optional[str] = None
    breedingMethodDbId: Optional[str] = None
    breedingMethodName: Optional[str] = None
    collection: Optional[str] = None
    commonCropName: str
    countryOfOriginCode: Optional[str] = None
    defaultDisplayName: Optional[str] = None
    documentationURL: Optional[str] = None
    genus: Optional[str] = None
    germplasmName: str
    germplasmPUI: str
    germplasmPreprocessing: Optional[str] = None
    instituteCode: Optional[str] = None
    instituteName: Optional[str] = None
    pedigree: Optional[str] = None
    sampleDbIds: Optional[List[str]] = None
    seedSource: Optional[str] = None
    seedSourceDescription: Optional[str] = None
    species: Optional[str] = None
    speciesAuthority: Optional[str] = None
    subtaxa: Optional[str] = None
    subtaxaAuthority: Optional[str] = None


class GermplasmAttribute(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    attributeCategory: Optional[str] = None
    attributeDescription: Optional[str] = None
    attributeName: str
    attributePUI: Optional[str] = None
    attributeValueDbIds: Optional[List[str]] = None
    commonCropName: Optional[str] = None
    defaultValue: Optional[str] = None
    documentationURL: Optional[str] = None
    growthStage: Optional[str] = None
    institution: Optional[str] = None
    language: Optional[str] = None
    methodDbId: Optional[str] = None
    methodName: str
    methodPUI: Optional[str] = None
    ontologyReferenceDbId: Optional[str] = None
    scaleDbId: str
    scaleName: str
    scalePUI: Optional[str] = None
    scientist: Optional[str] = None
    status: Optional[str] = None
    submissionTimestamp: Optional[str] = None
    traitDbId: Optional[str] = None
    traitName: str
    traitPUI: Optional[str] = None


class GermplasmAttributeNewRequest(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    attributeCategory: Optional[str] = None
    attributeDbId: str
    attributeDescription: Optional[str] = None
    attributeName: str
    attributePUI: Optional[str] = None
    attributeValueDbIds: Optional[List[str]] = None
    commonCropName: Optional[str] = None
    defaultValue: Optional[str] = None
    documentationURL: Optional[str] = None
    growthStage: Optional[str] = None
    institution: Optional[str] = None
    language: Optional[str] = None
    methodDbId: Optional[str] = None
    methodName: str
    methodPUI: Optional[str] = None
    ontologyReferenceDbId: Optional[str] = None
    scaleDbId: str
    scaleName: str
    scalePUI: Optional[str] = None
    scientist: Optional[str] = None
    status: Optional[str] = None
    submissionTimestamp: Optional[str] = None
    traitDbId: Optional[str] = None
    traitName: str
    traitPUI: Optional[str] = None


class GermplasmAttributeSearchRequest(BaseModel):
    pass


class GermplasmAttributeValue(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    attributeDbId: Optional[str] = None
    attributeName: str
    attributePUI: Optional[str] = None
    determinedDate: Optional[str] = None
    germplasmDbId: Optional[str] = None
    germplasmName: Optional[str] = None
    germplasmPUI: Optional[str] = None
    value: Optional[str] = None


class GermplasmAttributeValueNewRequest(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    attributeDbId: Optional[str] = None
    attributeName: str
    attributePUI: Optional[str] = None
    attributeValueDbId: str
    determinedDate: Optional[str] = None
    germplasmDbId: Optional[str] = None
    germplasmName: Optional[str] = None
    germplasmPUI: Optional[str] = None
    value: Optional[str] = None


class GermplasmAttributeValueSearchRequest(BaseModel):
    pass


class GermplasmNewRequest(BaseModel):
    accessionNumber: Optional[str] = None
    acquisitionDate: Optional[str] = None
    additionalInfo: Optional[AdditionalInfo] = None
    biologicalStatusOfAccessionCode: Optional[BiologicalStatusOfAccessionCode] = Field(
        None,
        description='MCPD (v2.1) (SAMPSTAT) 19. The coding scheme proposed can be used at 3 different levels of detail: either by using the general codes such as 100, 200, 300, 400, or by using the more specific codes such as 110, 120, etc. \n\n100) Wild \n110) Natural \n120) Semi-natural/wild \n130) Semi-natural/sown \n200) Weedy \n300) Traditional cultivar/landrace \n400) Breeding/research material \n410) Breeders line \n411) Synthetic population \n412) Hybrid \n413) Founder stock/base population \n414) Inbred line (parent of hybrid cultivar) \n415) Segregating population \n416) Clonal selection \n420) Genetic stock \n421) Mutant (e.g. induced/insertion mutants, tilling populations) \n422) Cytogenetic stocks (e.g. chromosome addition/substitution, aneuploids,  amphiploids) \n423) Other genetic stocks (e.g. mapping populations) \n500) Advanced or improved cultivar (conventional breeding methods) \n600) GMO (by genetic engineering) \n999) Other (Elaborate in REMARKS field)',
    )
    biologicalStatusOfAccessionDescription: Optional[str] = None
    breedingMethodDbId: Optional[str] = None
    breedingMethodName: Optional[str] = None
    collection: Optional[str] = None
    commonCropName: str
    countryOfOriginCode: Optional[str] = None
    defaultDisplayName: Optional[str] = None
    documentationURL: Optional[str] = None
    genus: Optional[str] = None
    germplasmDbId: str
    germplasmName: str
    germplasmPUI: str
    germplasmPreprocessing: Optional[str] = None
    instituteCode: Optional[str] = None
    instituteName: Optional[str] = None
    pedigree: Optional[str] = None
    sampleDbIds: Optional[List[str]] = None
    seedSource: Optional[str] = None
    seedSourceDescription: Optional[str] = None
    species: Optional[str] = None
    speciesAuthority: Optional[str] = None
    subtaxa: Optional[str] = None
    subtaxaAuthority: Optional[str] = None


class GermplasmOrigin(BaseModel):
    coordinateUncertainty: Optional[str] = None
    germplasmDbId: Optional[str] = None
    germplasmName: Optional[str] = None
    germplasmPUI: Optional[str] = None


class GermplasmSearchRequest(BaseModel):
    pass


class GrowthFacility(BaseModel):
    PUI: Optional[str] = None
    description: Optional[str] = None
    studyDbId: Optional[str] = None
    studyName: Optional[str] = None
    studyPUI: Optional[str] = None


class LastUpdate(BaseModel):
    lastUpdateDbId: str
    studyDbId: Optional[str] = None
    studyName: Optional[str] = None
    studyPUI: Optional[str] = None
    timestamp: Optional[str] = None
    version: Optional[str] = None


class ListType(Enum):
    germplasm = 'germplasm'
    markers = 'markers'
    variants = 'variants'
    programs = 'programs'
    trials = 'trials'
    studies = 'studies'
    observationUnits = 'observationUnits'
    observations = 'observations'
    observationVariables = 'observationVariables'
    samples = 'samples'


class MarkerPosition(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    linkageGroupName: Optional[str] = None
    mapDbId: Optional[str] = None
    mapName: Optional[str] = None
    mapPUI: Optional[str] = None
    position: Optional[int] = None
    variantDbId: Optional[str] = None


class MarkerPositionSearchRequest(BaseModel):
    maxPosition: Optional[int] = None
    minPosition: Optional[int] = None


class MetadataField(BaseModel):
    dataType: Optional[DataType] = Field(
        None,
        description='The type of field represented in this Genotype Field. This is intended to help parse the data out of JSON.',
    )
    fieldAbbreviation: Optional[str] = None
    fieldName: Optional[str] = None
    variantSetDbId: Optional[str] = None
    variantSetName: Optional[str] = None


class Method(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    bibliographicalReference: Optional[str] = None
    description: Optional[str] = None
    formula: Optional[str] = None
    methodClass: Optional[str] = None
    methodName: str
    methodPUI: Optional[str] = None
    ontologyReferenceDbId: Optional[str] = None


class MethodNewRequest(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    bibliographicalReference: Optional[str] = None
    description: Optional[str] = None
    formula: Optional[str] = None
    methodClass: Optional[str] = None
    methodDbId: str
    methodName: str
    methodPUI: Optional[str] = None
    ontologyReferenceDbId: Optional[str] = None


class ObservationSearchRequest(BaseModel):
    observationTimeStampRangeEnd: Optional[str] = None
    observationTimeStampRangeStart: Optional[str] = None


class ObservationUnit(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    crossDbId: Optional[str] = None
    crossName: Optional[str] = None
    germplasmDbId: Optional[str] = None
    germplasmName: Optional[str] = None
    germplasmPUI: Optional[str] = None
    locationDbId: Optional[str] = None
    locationName: Optional[str] = None
    observationUnitName: Optional[str] = None
    observationUnitPUI: Optional[str] = None
    programDbId: Optional[str] = None
    programName: Optional[str] = None
    seedLotDbId: Optional[str] = None
    seedLotName: Optional[str] = None
    studyDbId: Optional[str] = None
    studyName: Optional[str] = None
    studyPUI: Optional[str] = None
    trialDbId: Optional[str] = None
    trialName: Optional[str] = None
    trialPUI: Optional[str] = None


class ObservationUnitHierarchyLevel(BaseModel):
    levelName: Optional[str] = None
    levelOrder: Optional[int] = None


class ObservationUnitLevel(BaseModel):
    levelCode: Optional[str] = None
    levelName: Optional[str] = None
    levelOrder: Optional[int] = None


class ObservationUnitLevelRelationship(BaseModel):
    levelCode: Optional[str] = None
    levelName: Optional[str] = None
    levelOrder: Optional[int] = None
    observationUnit: Optional[ObservationUnit] = None


class ObservationUnitNewRequest(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    crossDbId: Optional[str] = None
    crossName: Optional[str] = None
    germplasmDbId: Optional[str] = None
    germplasmName: Optional[str] = None
    germplasmPUI: Optional[str] = None
    locationDbId: Optional[str] = None
    locationName: Optional[str] = None
    observationUnitDbId: str
    observationUnitName: Optional[str] = None
    observationUnitPUI: Optional[str] = None
    programDbId: Optional[str] = None
    programName: Optional[str] = None
    seedLotDbId: Optional[str] = None
    seedLotName: Optional[str] = None
    studyDbId: Optional[str] = None
    studyName: Optional[str] = None
    studyPUI: Optional[str] = None
    trialDbId: Optional[str] = None
    trialName: Optional[str] = None
    trialPUI: Optional[str] = None


class EntryType(Enum):
    CHECK = 'CHECK'
    TEST = 'TEST'
    FILLER = 'FILLER'


class PositionCoordinateXType(Enum):
    LONGITUDE = 'LONGITUDE'
    LATITUDE = 'LATITUDE'
    PLANTED_ROW = 'PLANTED_ROW'
    PLANTED_INDIVIDUAL = 'PLANTED_INDIVIDUAL'
    GRID_ROW = 'GRID_ROW'
    GRID_COL = 'GRID_COL'
    MEASURED_ROW = 'MEASURED_ROW'
    MEASURED_COL = 'MEASURED_COL'


class PositionCoordinateYType(Enum):
    LONGITUDE = 'LONGITUDE'
    LATITUDE = 'LATITUDE'
    PLANTED_ROW = 'PLANTED_ROW'
    PLANTED_INDIVIDUAL = 'PLANTED_INDIVIDUAL'
    GRID_ROW = 'GRID_ROW'
    GRID_COL = 'GRID_COL'
    MEASURED_ROW = 'MEASURED_ROW'
    MEASURED_COL = 'MEASURED_COL'


class ObservationUnitPosition(BaseModel):
    entryType: Optional[EntryType] = Field(
        None,
        description='The type of entry for this observation unit. ex. "CHECK", "TEST", "FILLER"',
    )
    observationLevel: Optional[ObservationUnitLevel] = None
    observationUnitDbId: Optional[str] = None
    observationUnitName: Optional[str] = None
    observationUnitPUI: Optional[str] = None
    positionCoordinateX: Optional[str] = None
    positionCoordinateXType: Optional[PositionCoordinateXType] = Field(
        None,
        description="The type of positional coordinate used. Must be one of the following values \n\nLONGITUDE - ISO 6709 standard, WGS84 geodetic datum. See 'Location Coordinate Encoding' for details \n\nLATITUDE - ISO 6709 standard, WGS84 geodetic datum. See 'Location Coordinate Encoding' for details \n\nPLANTED_ROW - The physical planted row number \n\nPLANTED_INDIVIDUAL - The physical counted number, could be independent or within a planted row \n\nGRID_ROW - The row index number of a square grid overlay \n\nGRID_COL - The column index number of a square grid overlay \n\nMEASURED_ROW - The distance in meters from a defined 0-th row \n\nMEASURED_COL - The distance in meters from a defined 0-th column ",
    )
    positionCoordinateY: Optional[str] = None
    positionCoordinateYType: Optional[PositionCoordinateYType] = Field(
        None,
        description="The type of positional coordinate used. Must be one of the following values \n\nLONGITUDE - ISO 6709 standard, WGS84 geodetic datum. See 'Location Coordinate Encoding' for details \n\nLATITUDE - ISO 6709 standard, WGS84 geodetic datum. See 'Location Coordinate Encoding' for details \n\nPLANTED_ROW - The physical planted row number  \n\nPLANTED_INDIVIDUAL - The physical counted number, could be independent or within a planted row \n\nGRID_ROW - The row index number of a square grid overlay \n\nGRID_COL - The column index number of a square grid overlay \n\nMEASURED_ROW - The distance in meters from a defined 0-th row \n\nMEASURED_COL - The distance in meters from a defined 0-th column ",
    )


class ObservationUnitSearchRequest(BaseModel):
    includeObservations: Optional[bool] = None


class ObservationVariable(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    commonCropName: Optional[str] = None
    defaultValue: Optional[str] = None
    documentationURL: Optional[str] = None
    growthStage: Optional[str] = None
    institution: Optional[str] = None
    language: Optional[str] = None
    methodDbId: Optional[str] = None
    methodName: str
    methodPUI: Optional[str] = None
    observationVariableName: str
    observationVariablePUI: Optional[str] = None
    ontologyReferenceDbId: Optional[str] = None
    scaleDbId: str
    scaleName: str
    scalePUI: Optional[str] = None
    scientist: Optional[str] = None
    status: Optional[str] = None
    submissionTimestamp: Optional[str] = None
    traitDbId: Optional[str] = None
    traitName: str
    traitPUI: Optional[str] = None


class ObservationVariableNewRequest(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    commonCropName: Optional[str] = None
    defaultValue: Optional[str] = None
    documentationURL: Optional[str] = None
    growthStage: Optional[str] = None
    institution: Optional[str] = None
    language: Optional[str] = None
    methodDbId: Optional[str] = None
    methodName: str
    methodPUI: Optional[str] = None
    observationVariableDbId: str
    observationVariableName: str
    observationVariablePUI: Optional[str] = None
    ontologyReferenceDbId: Optional[str] = None
    scaleDbId: str
    scaleName: str
    scalePUI: Optional[str] = None
    scientist: Optional[str] = None
    status: Optional[str] = None
    submissionTimestamp: Optional[str] = None
    traitDbId: Optional[str] = None
    traitName: str
    traitPUI: Optional[str] = None


class ObservationVariableSearchRequest(BaseModel):
    pass


class Ontology(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    authors: Optional[str] = None
    copyright: Optional[str] = None
    description: Optional[str] = None
    documentationURL: Optional[str] = None
    licence: Optional[str] = None
    ontologyName: str
    version: Optional[str] = None


class OntologyNewRequest(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    authors: Optional[str] = None
    copyright: Optional[str] = None
    description: Optional[str] = None
    documentationURL: Optional[str] = None
    licence: Optional[str] = None
    ontologyDbId: str
    ontologyName: str
    version: Optional[str] = None


class OntologyReference(BaseModel):
    ontology: Ontology
    ontologyReferenceDbId: Optional[str] = None
    version: Optional[str] = None


class OntologyTerm(BaseModel):
    term: Optional[str] = None
    termURI: Optional[str] = None


class Dimension(Enum):
    CALLSETS = 'CALLSETS'
    VARIANTS = 'VARIANTS'


class Pagination(BaseModel):
    alleleMatrix: Optional[AlleleMatrix] = None
    dimension: Optional[Dimension] = Field(
        None, description='The dimension of the matrix being paginated'
    )
    page: Optional[int] = None
    pageSize: Optional[int] = None
    totalCount: Optional[int] = None
    totalPages: Optional[int] = None


class ParentType(Enum):
    MALE = 'MALE'
    FEMALE = 'FEMALE'
    SELF = 'SELF'
    POPULATION = 'POPULATION'
    CLONAL = 'CLONAL'


class PedigreeNode(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    breedingMethodDbId: Optional[str] = None
    breedingMethodName: Optional[str] = None
    crossingProjectDbId: Optional[str] = None
    crossingProjectName: Optional[str] = None
    crossingYear: Optional[int] = None
    defaultDisplayName: Optional[str] = None
    familyCode: Optional[str] = None
    germplasmDbId: str
    germplasmName: str
    germplasmPUI: str
    pedigreeNodeDbId: Optional[str] = None
    pedigreeString: Optional[str] = None
    siblingDbIds: Optional[List[str]] = None


class PedigreeNodeSearchRequest(BaseModel):
    includeFullTree: Optional[bool] = None
    includeParents: Optional[bool] = None
    includeProgeny: Optional[bool] = None
    includeSiblings: Optional[bool] = None
    pedigreeDepth: Optional[int] = None
    progenyDepth: Optional[int] = None


class Person(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    description: Optional[str] = None
    emailAddress: Optional[str] = None
    firstName: Optional[str] = None
    lastName: Optional[str] = None
    mailingAddress: Optional[str] = None
    middleName: Optional[str] = None
    phoneNumber: Optional[str] = None
    userID: Optional[str] = None


class PersonNewRequest(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    description: Optional[str] = None
    emailAddress: Optional[str] = None
    firstName: Optional[str] = None
    lastName: Optional[str] = None
    mailingAddress: Optional[str] = None
    middleName: Optional[str] = None
    personDbId: str
    phoneNumber: Optional[str] = None
    userID: Optional[str] = None


class PersonSearchRequest(BaseModel):
    pass


class Status(Enum):
    TODO = 'TODO'
    DONE = 'DONE'
    SKIPPED = 'SKIPPED'


class PlateFormat(Enum):
    PLATE_96 = 'PLATE_96'
    TUBES = 'TUBES'


class SampleType(Enum):
    DNA = 'DNA'
    RNA = 'RNA'
    TISSUE = 'TISSUE'
    MIXED = 'MIXED'


class Plate(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    plateBarcode: Optional[str] = None
    plateFormat: Optional[PlateFormat] = Field(
        None,
        description='Enum for plate formats, usually "PLATE_96" for a 96 well plate or "TUBES" for plateless format',
    )
    plateName: str
    programDbId: Optional[str] = None
    programName: Optional[str] = None
    sampleType: Optional[SampleType] = Field(
        None, description="The type of samples taken. ex. 'DNA', 'RNA', 'Tissue', etc"
    )
    studyDbId: Optional[str] = None
    studyName: Optional[str] = None
    studyPUI: Optional[str] = None
    trialDbId: Optional[str] = None
    trialName: Optional[str] = None
    trialPUI: Optional[str] = None


class PlateNewRequest(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    plateBarcode: Optional[str] = None
    plateDbId: str
    plateFormat: Optional[PlateFormat] = Field(
        None,
        description='Enum for plate formats, usually "PLATE_96" for a 96 well plate or "TUBES" for plateless format',
    )
    plateName: str
    programDbId: Optional[str] = None
    programName: Optional[str] = None
    sampleType: Optional[SampleType] = Field(
        None, description="The type of samples taken. ex. 'DNA', 'RNA', 'Tissue', etc"
    )
    studyDbId: Optional[str] = None
    studyName: Optional[str] = None
    studyPUI: Optional[str] = None
    trialDbId: Optional[str] = None
    trialName: Optional[str] = None
    trialPUI: Optional[str] = None


class PlateSearchRequest(BaseModel):
    pass


class PollinationEvent(BaseModel):
    crossDbId: Optional[str] = None
    crossName: Optional[str] = None
    pollinationNumber: Optional[str] = None
    pollinationSuccessful: Optional[bool] = None
    pollinationTimeStamp: Optional[str] = None


class ProgramType(Enum):
    STANDARD = 'STANDARD'
    PROJECT = 'PROJECT'


class Program(BaseModel):
    abbreviation: Optional[str] = None
    additionalInfo: Optional[AdditionalInfo] = None
    commonCropName: Optional[str] = None
    documentationURL: Optional[str] = None
    fundingInformation: Optional[str] = None
    objective: Optional[str] = None
    personDbId: Optional[str] = None
    programName: str
    programType: Optional[ProgramType] = Field(
        None,
        description="The type of program entity this object represents\n<br/> 'STANDARD' represents a standard, permanent breeding program\n<br/> 'PROJECT' represents a short term project, usually with a set time limit based on funding   ",
    )


class ProgramNewRequest(BaseModel):
    abbreviation: Optional[str] = None
    additionalInfo: Optional[AdditionalInfo] = None
    commonCropName: Optional[str] = None
    documentationURL: Optional[str] = None
    fundingInformation: Optional[str] = None
    objective: Optional[str] = None
    personDbId: Optional[str] = None
    programDbId: str
    programName: str
    programType: Optional[ProgramType] = Field(
        None,
        description="The type of program entity this object represents\n<br/> 'STANDARD' represents a standard, permanent breeding program\n<br/> 'PROJECT' represents a short term project, usually with a set time limit based on funding   ",
    )


class ProgramSearchRequest(BaseModel):
    pass


class Publication(BaseModel):
    publicationPUI: Optional[str] = None
    publicationReference: Optional[str] = None
    trialDbId: Optional[str] = None
    trialName: Optional[str] = None
    trialPUI: Optional[str] = None


class Reference(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    commonCropName: Optional[str] = None
    isDerived: Optional[bool] = None
    length: Optional[int] = None
    md5checksum: Optional[str] = None
    referenceDbId: str
    referenceName: str
    referenceSetDbId: Optional[str] = None
    referenceSetName: Optional[str] = None
    sourceDivergence: Optional[float] = None
    sourceGermplasmDbIds: Optional[List[str]] = None
    sourceURI: Optional[str] = None
    species: Optional[OntologyTerm] = None
    variantDbIds: Optional[List[str]] = None


class ReferenceSearchRequest(BaseModel):
    isDerived: Optional[bool] = None
    maxLength: Optional[int] = None
    minLength: Optional[int] = None


class ReferenceSet(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    assemblyPUI: Optional[str] = None
    commonCropName: Optional[str] = None
    description: Optional[str] = None
    isDerived: Optional[bool] = None
    md5checksum: Optional[str] = None
    referenceDbId: Optional[str] = None
    referenceName: Optional[str] = None
    referenceSetName: str
    sourceGermplasmDbIds: Optional[List[str]] = None
    sourceURI: Optional[str] = None
    species: Optional[OntologyTerm] = None
    variantDbIds: Optional[List[str]] = None
    variantSetDbIds: Optional[List[str]] = None


class ReferenceSetNewRequest(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    assemblyPUI: Optional[str] = None
    commonCropName: Optional[str] = None
    description: Optional[str] = None
    isDerived: Optional[bool] = None
    md5checksum: Optional[str] = None
    referenceDbId: Optional[str] = None
    referenceName: Optional[str] = None
    referenceSetDbId: str
    referenceSetName: str
    sourceGermplasmDbIds: Optional[List[str]] = None
    sourceURI: Optional[str] = None
    species: Optional[OntologyTerm] = None
    variantDbIds: Optional[List[str]] = None
    variantSetDbIds: Optional[List[str]] = None


class ReferenceSetSearchRequest(BaseModel):
    pass


class Sample(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    callSetDbIds: Optional[List[str]] = None
    column: Optional[int] = None
    germplasmDbId: Optional[str] = None
    germplasmName: Optional[str] = None
    germplasmPUI: Optional[str] = None
    observationUnitDbId: Optional[str] = None
    observationUnitName: Optional[str] = None
    observationUnitPUI: Optional[str] = None
    plateDbId: Optional[str] = None
    plateName: Optional[str] = None
    programDbId: Optional[str] = None
    programName: Optional[str] = None
    row: Optional[str] = None
    sampleBarcode: Optional[str] = None
    sampleDescription: Optional[str] = None
    sampleGroupId: Optional[str] = None
    sampleName: str
    samplePUI: Optional[str] = None
    sampleTimestamp: Optional[str] = None
    sampleType: Optional[str] = None
    studyDbId: Optional[str] = None
    studyName: Optional[str] = None
    studyPUI: Optional[str] = None
    takenBy: Optional[str] = None
    tissueType: Optional[str] = None
    trialDbId: Optional[str] = None
    trialName: Optional[str] = None
    trialPUI: Optional[str] = None
    well: Optional[str] = None


class SampleNewRequest(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    callSetDbIds: Optional[List[str]] = None
    column: Optional[int] = None
    germplasmDbId: Optional[str] = None
    germplasmName: Optional[str] = None
    germplasmPUI: Optional[str] = None
    observationUnitDbId: Optional[str] = None
    observationUnitName: Optional[str] = None
    observationUnitPUI: Optional[str] = None
    plateDbId: Optional[str] = None
    plateName: Optional[str] = None
    programDbId: Optional[str] = None
    programName: Optional[str] = None
    row: Optional[str] = None
    sampleBarcode: Optional[str] = None
    sampleDbId: str
    sampleDescription: Optional[str] = None
    sampleGroupId: Optional[str] = None
    sampleName: str
    samplePUI: Optional[str] = None
    sampleTimestamp: Optional[str] = None
    sampleType: Optional[str] = None
    studyDbId: Optional[str] = None
    studyName: Optional[str] = None
    studyPUI: Optional[str] = None
    takenBy: Optional[str] = None
    tissueType: Optional[str] = None
    trialDbId: Optional[str] = None
    trialName: Optional[str] = None
    trialPUI: Optional[str] = None
    well: Optional[str] = None


class SampleSearchRequest(BaseModel):
    pass


class DataType3(Enum):
    Code = 'Code'
    Date = 'Date'
    Duration = 'Duration'
    Nominal = 'Nominal'
    Numerical = 'Numerical'
    Ordinal = 'Ordinal'
    Text = 'Text'


class Season(BaseModel):
    seasonDbId: str
    seasonName: Optional[str] = None
    year: Optional[int] = None


class SeedLot(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    amount: Optional[float] = None
    createdDate: Optional[str] = None
    lastUpdated: Optional[str] = None
    locationDbId: Optional[str] = None
    locationName: Optional[str] = None
    programDbId: Optional[str] = None
    programName: Optional[str] = None
    seedLotDescription: Optional[str] = None
    seedLotName: str
    sourceCollection: Optional[str] = None
    storageLocation: Optional[str] = None
    units: Optional[str] = None


class SeedLotNewRequest(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    amount: Optional[float] = None
    createdDate: Optional[str] = None
    lastUpdated: Optional[str] = None
    locationDbId: Optional[str] = None
    locationName: Optional[str] = None
    programDbId: Optional[str] = None
    programName: Optional[str] = None
    seedLotDbId: str
    seedLotDescription: Optional[str] = None
    seedLotName: str
    sourceCollection: Optional[str] = None
    storageLocation: Optional[str] = None
    units: Optional[str] = None


class SeedLotTransaction(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    amount: Optional[float] = None
    seedLotDbId: Optional[str] = None
    seedLotName: Optional[str] = None
    transactionDbId: str
    transactionDescription: Optional[str] = None
    transactionTimestamp: Optional[str] = None
    units: Optional[str] = None


class Code(Enum):
    field_10 = '10'
    field_11 = '11'
    field_12 = '12'
    field_13 = '13'
    field_20 = '20'
    field_30 = '30'
    field_40 = '40'
    field_50 = '50'
    field_99 = '99'


class StorageType(BaseModel):
    code: Optional[Code] = Field(
        None,
        description='The 2 digit code representing the type of storage this germplasm is kept in at a genebank. \n\nMCPD (v2.1) (STORAGE) 26. If germplasm is maintained under different types of storage, multiple choices are allowed, separated by a semicolon (e.g. 20;30). (Refer to FAO/IPGRI Genebank Standards 1994 for details on storage type.) \n\n10) Seed collection \n11) Short term \n12) Medium term \n13) Long term \n20) Field collection \n30) In vitro collection \n40) Cryo-preserved collection \n50) DNA collection \n99) Other (elaborate in REMARKS field)',
    )
    description: Optional[str] = None
    germplasmDbId: Optional[str] = None
    germplasmName: Optional[str] = None
    germplasmPUI: Optional[str] = None


class Study(BaseModel):
    active: Optional[bool] = None
    additionalInfo: Optional[AdditionalInfo] = None
    commonCropName: Optional[str] = None
    contactDbIds: Optional[List[str]] = None
    culturalPractices: Optional[str] = None
    documentationURL: Optional[str] = None
    endDate: Optional[str] = None
    environmentParameterDbIds: Optional[List[str]] = None
    experimentalDesign: Optional[ExperimentalDesign] = None
    growthFacility: Optional[GrowthFacility] = None
    lastUpdateDbId: Optional[str] = None
    license: Optional[str] = None
    locationDbId: Optional[str] = None
    locationName: Optional[str] = None
    observationUnitsDescription: Optional[str] = None
    observationVariableDbIds: Optional[List[str]] = None
    startDate: Optional[str] = None
    studyCode: Optional[str] = None
    studyDescription: Optional[str] = None
    studyName: str
    studyPUI: Optional[str] = None
    studyType: Optional[str] = None
    trialDbId: Optional[str] = None
    trialName: Optional[str] = None
    trialPUI: Optional[str] = None


class StudyNewRequest(BaseModel):
    active: Optional[bool] = None
    additionalInfo: Optional[AdditionalInfo] = None
    commonCropName: Optional[str] = None
    contactDbIds: Optional[List[str]] = None
    culturalPractices: Optional[str] = None
    documentationURL: Optional[str] = None
    endDate: Optional[str] = None
    environmentParameterDbIds: Optional[List[str]] = None
    experimentalDesign: Optional[ExperimentalDesign] = None
    growthFacility: Optional[GrowthFacility] = None
    lastUpdateDbId: Optional[str] = None
    license: Optional[str] = None
    locationDbId: Optional[str] = None
    locationName: Optional[str] = None
    observationUnitsDescription: Optional[str] = None
    observationVariableDbIds: Optional[List[str]] = None
    startDate: Optional[str] = None
    studyCode: Optional[str] = None
    studyDbId: str
    studyDescription: Optional[str] = None
    studyName: str
    studyPUI: Optional[str] = None
    studyType: Optional[str] = None
    trialDbId: Optional[str] = None
    trialName: Optional[str] = None
    trialPUI: Optional[str] = None


class StudySearchRequest(BaseModel):
    active: Optional[bool] = None


class Synonym(BaseModel):
    germplasmDbId: Optional[str] = None
    germplasmName: Optional[str] = None
    germplasmPUI: Optional[str] = None
    synonym: Optional[str] = None
    type: Optional[str] = None


class TaxonId(BaseModel):
    germplasmDbId: Optional[str] = None
    germplasmName: Optional[str] = None
    germplasmPUI: Optional[str] = None
    sourceName: str
    taxonId: str


class Trait(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    attribute: Optional[str] = None
    attributePUI: Optional[str] = None
    entity: Optional[str] = None
    entityPUI: Optional[str] = None
    mainAbbreviation: Optional[str] = None
    ontologyReferenceDbId: Optional[str] = None
    status: Optional[str] = None
    traitClass: Optional[str] = None
    traitDescription: Optional[str] = None
    traitName: str
    traitPUI: Optional[str] = None


class TraitDataType(Enum):
    Code = 'Code'
    Date = 'Date'
    Duration = 'Duration'
    Nominal = 'Nominal'
    Numerical = 'Numerical'
    Ordinal = 'Ordinal'
    Text = 'Text'


class TraitNewRequest(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    attribute: Optional[str] = None
    attributePUI: Optional[str] = None
    entity: Optional[str] = None
    entityPUI: Optional[str] = None
    mainAbbreviation: Optional[str] = None
    ontologyReferenceDbId: Optional[str] = None
    status: Optional[str] = None
    traitClass: Optional[str] = None
    traitDbId: str
    traitDescription: Optional[str] = None
    traitName: str
    traitPUI: Optional[str] = None


class Treatment(BaseModel):
    factor: Optional[str] = None
    modality: Optional[str] = None
    observationUnitDbId: Optional[str] = None
    observationUnitName: Optional[str] = None
    observationUnitPUI: Optional[str] = None


class Trial(BaseModel):
    active: Optional[bool] = None
    additionalInfo: Optional[AdditionalInfo] = None
    commonCropName: Optional[str] = None
    contacts: Optional[List[Contact]] = None
    documentationURL: Optional[str] = None
    endDate: Optional[str] = None
    programDbId: Optional[str] = None
    programName: Optional[str] = None
    startDate: Optional[str] = None
    studyDbIds: Optional[List[str]] = None
    trialDescription: Optional[str] = None
    trialName: str
    trialPUI: Optional[str] = None


class TrialNewRequest(BaseModel):
    active: Optional[bool] = None
    additionalInfo: Optional[AdditionalInfo] = None
    commonCropName: Optional[str] = None
    contacts: Optional[List[Contact]] = None
    documentationURL: Optional[str] = None
    endDate: Optional[str] = None
    programDbId: Optional[str] = None
    programName: Optional[str] = None
    startDate: Optional[str] = None
    studyDbIds: Optional[List[str]] = None
    trialDbId: str
    trialDescription: Optional[str] = None
    trialName: str
    trialPUI: Optional[str] = None


class TrialSearchRequest(BaseModel):
    active: Optional[bool] = None
    searchDateRangeEnd: Optional[str] = None
    searchDateRangeStart: Optional[str] = None


class ValidValues(BaseModel):
    maximumValue: Optional[str] = None
    minimumValue: Optional[str] = None
    scaleDbId: Optional[str] = None
    scaleName: Optional[str] = None
    scalePUI: Optional[str] = None


class ValidValuesCategory(BaseModel):
    label: Optional[str] = None
    validValues: Optional[ValidValues] = None
    value: Optional[str] = None


class Variant(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    created: Optional[str] = None
    end: Optional[int] = None
    filtersApplied: Optional[bool] = None
    filtersPassed: Optional[bool] = None
    referenceBases: Optional[str] = None
    referenceDbId: Optional[str] = None
    referenceName: Optional[str] = None
    referenceSetDbId: Optional[str] = None
    referenceSetName: Optional[str] = None
    start: Optional[int] = None
    svlen: Optional[int] = None
    updated: Optional[str] = None
    variantDbId: str
    variantSetDbId: Optional[str] = None
    variantSetName: Optional[str] = None
    variantType: Optional[str] = None


class VariantSearchRequest(BaseModel):
    end: Optional[int] = None
    referenceDbId: Optional[str] = None
    start: Optional[int] = None


class VariantSet(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    analysiDbIds: Optional[List[str]] = None
    callSetCount: Optional[int] = None
    referenceSetDbId: Optional[str] = None
    referenceSetName: Optional[str] = None
    studyDbId: Optional[str] = None
    studyName: Optional[str] = None
    studyPUI: Optional[str] = None
    variantCount: Optional[int] = None
    variantSetDbId: str
    variantSetName: Optional[str] = None


class VariantSetSearchRequest(BaseModel):
    pass


class BasePagination(BaseModel):
    currentPage: int = Field(
        ...,
        description='The index number for the returned page of data. This should always match the requested page number or the default page (0).',
        examples=[0],
    )
    pageSize: int = Field(
        ...,
        description='The number of data elements returned, aka the size of the current page. If the requested page does not have enough elements to fill a page at the requested page size, this field should indicate the actual number of elements returned.',
        examples=[1000],
    )
    totalCount: Optional[int] = Field(
        None,
        description='The total number of elements that are available on the ca and match the requested query parameters.',
        examples=[10],
    )
    totalPages: Optional[int] = Field(
        None,
        description='The total number of pages of elements available on the server. This should be calculated with the following formula. \n<br> totalPages = CEILING( totalCount / requested_page_size)',
        examples=[1],
    )


class DataFile(BaseModel):
    fileURL: AnyUrl = Field(
        ...,
        description='The absolute URL where the file is located',
        examples=['https://wiki.brapi.org/examples/datafile.xlsx'],
    )
    fileName: Optional[str] = Field(
        None, description='The name of the file', examples=['datafile.xlsx']
    )
    fileSize: Optional[int] = Field(
        None, description='The size of the file in bytes', examples=[4398]
    )
    fileDescription: Optional[str] = Field(
        None,
        description='A human readable description of the file contents',
        examples=['This is an Excel data file'],
    )
    fileType: Optional[str] = Field(
        None,
        description='The type or format of the file. Preferably MIME Type.',
        examples=['application/vnd.ms-excel'],
    )
    fileMD5Hash: Optional[str] = Field(
        None,
        description='The MD5 Hash of the file contents to be used as a check sum',
        examples=['c2365e900c81a89cf74d83dab60df146'],
    )


class MessageType(Enum):
    DEBUG = 'DEBUG'
    ERROR = 'ERROR'
    WARNING = 'WARNING'
    INFO = 'INFO'


class StatusModel(BaseModel):
    message: str = Field(
        ...,
        description='A short message concerning the status of this request/response',
        examples=['Request accepted, response successful'],
    )
    messageType: MessageType = Field(
        ..., description='The logging level for the attached message', examples=['INFO']
    )


class TokenPagination(BasePagination):
    nextPageToken: str = Field(
        ...,
        description='**Deprecated in v2.1** Please use `page`. Github issue number #451 \n<br>The string token used to query the next page of data.',
        examples=['cb668f63'],
    )
    currentPageToken: Optional[str] = Field(
        None,
        description='**Deprecated in v2.1** Please use `page`. Github issue number #451 \n<br>The string token used to query the current page of data.',
        examples=['48bc6ac1'],
    )
    prevPageToken: Optional[str] = Field(
        None,
        description='**Deprecated in v2.1** Please use `page`. Github issue number #451 \n<br>The string token used to query the previous page of data.',
        examples=['9659857e'],
    )


class CrossParent(BaseModel):
    germplasm: Optional[Germplasm] = None
    observationUnitDbId: Optional[str] = None
    observationUnitName: Optional[str] = None
    observationUnitPUI: Optional[str] = None
    parentType: Optional[ParentType] = None


class GeoJSON(BaseModel):
    geometry: Optional[GeoJSONGeometry] = None
    type: Optional[str] = None


class GeoJSONSearchArea(BaseModel):
    geometry: Optional[GeoJSONGeometry] = None
    germplasmOrigin: Optional[GermplasmOrigin] = None
    imageDbId: Optional[str] = None
    imageName: Optional[str] = None
    observationDbId: Optional[str] = None
    observationUnit: Optional[ObservationUnitPosition] = None
    type: Optional[str] = None


class GermplasmChild(BaseModel):
    germplasmDbId: str
    germplasmName: str
    germplasmPUI: str
    parentType: ParentType
    pedigreeNodeDbId: Optional[str] = None


class GermplasmParent(BaseModel):
    germplasmDbId: str
    germplasmName: str
    germplasmPUI: str
    parentType: ParentType
    pedigreeNodeDbId: Optional[str] = None


class Image(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    copyright: Optional[str] = None
    description: Optional[str] = None
    imageFileName: Optional[str] = None
    imageFileSize: Optional[int] = None
    imageHeight: Optional[int] = None
    imageLocation: Optional[GeoJSON] = None
    imageName: Optional[str] = None
    imageTimeStamp: Optional[str] = None
    imageURL: Optional[str] = None
    imageWidth: Optional[int] = None
    mimeType: Optional[str] = None
    observationDbIds: Optional[List[str]] = None
    observationUnitDbId: Optional[str] = None
    observationUnitName: Optional[str] = None
    observationUnitPUI: Optional[str] = None


class ImageNewRequest(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    copyright: Optional[str] = None
    description: Optional[str] = None
    imageDbId: str
    imageFileName: Optional[str] = None
    imageFileSize: Optional[int] = None
    imageHeight: Optional[int] = None
    imageLocation: Optional[GeoJSON] = None
    imageName: Optional[str] = None
    imageTimeStamp: Optional[str] = None
    imageURL: Optional[str] = None
    imageWidth: Optional[int] = None
    mimeType: Optional[str] = None
    observationDbIds: Optional[List[str]] = None
    observationUnitDbId: Optional[str] = None
    observationUnitName: Optional[str] = None
    observationUnitPUI: Optional[str] = None


class ImageSearchRequest(BaseModel):
    imageFileSizeMax: Optional[int] = None
    imageFileSizeMin: Optional[int] = None
    imageHeightMax: Optional[int] = None
    imageHeightMin: Optional[int] = None
    imageLocation: Optional[GeoJSONSearchArea] = None
    imageTimeStampRangeEnd: Optional[str] = None
    imageTimeStampRangeStart: Optional[str] = None
    imageWidthMax: Optional[int] = None
    imageWidthMin: Optional[int] = None


class ListModel(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    dateCreated: Optional[str] = None
    dateModified: Optional[str] = None
    listDescription: Optional[str] = None
    listName: str
    listOwnerName: Optional[str] = None
    listSize: Optional[int] = None
    listSource: Optional[str] = None
    listType: ListType
    personDbId: Optional[str] = None


class ListNewRequest(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    dateCreated: Optional[str] = None
    dateModified: Optional[str] = None
    listDbId: str
    listDescription: Optional[str] = None
    listName: str
    listOwnerName: Optional[str] = None
    listSize: Optional[int] = None
    listSource: Optional[str] = None
    listType: ListType
    personDbId: Optional[str] = None


class ListSearchRequest(BaseModel):
    dateCreatedRangeEnd: Optional[str] = None
    dateCreatedRangeStart: Optional[str] = None
    dateModifiedRangeEnd: Optional[str] = None
    dateModifiedRangeStart: Optional[str] = None
    listType: Optional[ListType] = None


class Location(BaseModel):
    abbreviation: Optional[str] = None
    additionalInfo: Optional[AdditionalInfo] = None
    childLocationDbIds: Optional[List[str]] = None
    coordinateDescription: Optional[str] = None
    coordinateUncertainty: Optional[str] = None
    coordinates: Optional[GeoJSON] = None
    countryCode: Optional[str] = None
    countryName: Optional[str] = None
    documentationURL: Optional[str] = None
    environmentType: Optional[str] = None
    exposure: Optional[str] = None
    instituteAddress: Optional[str] = None
    instituteName: Optional[str] = None
    locationDbId: str
    locationName: str
    locationType: Optional[str] = None
    siteStatus: Optional[str] = None
    slope: Optional[str] = None
    topography: Optional[str] = None


class LocationNewRequest(BaseModel):
    abbreviation: Optional[str] = None
    additionalInfo: Optional[AdditionalInfo] = None
    childLocationDbIds: Optional[List[str]] = None
    coordinateDescription: Optional[str] = None
    coordinateUncertainty: Optional[str] = None
    coordinates: Optional[GeoJSON] = None
    countryCode: Optional[str] = None
    countryName: Optional[str] = None
    documentationURL: Optional[str] = None
    environmentType: Optional[str] = None
    exposure: Optional[str] = None
    instituteAddress: Optional[str] = None
    instituteName: Optional[str] = None
    locationDbId: str
    locationName: str
    locationType: Optional[str] = None
    siteStatus: Optional[str] = None
    slope: Optional[str] = None
    topography: Optional[str] = None


class LocationSearchRequest(BaseModel):
    altitudeMax: Optional[float] = None
    altitudeMin: Optional[float] = None
    coordinates: Optional[GeoJSONSearchArea] = None


class Observation(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    collector: Optional[str] = None
    geoCoordinates: Optional[GeoJSON] = None
    germplasmDbId: Optional[str] = None
    germplasmName: Optional[str] = None
    germplasmPUI: Optional[str] = None
    observationTimeStamp: Optional[str] = None
    observationUnitDbId: Optional[str] = None
    observationUnitName: Optional[str] = None
    observationUnitPUI: Optional[str] = None
    observationVariableDbId: Optional[str] = None
    observationVariableName: Optional[str] = None
    observationVariablePUI: Optional[str] = None
    seasonDbId: Optional[str] = None
    seasonName: Optional[str] = None
    studyDbId: Optional[str] = None
    studyName: Optional[str] = None
    studyPUI: Optional[str] = None
    uploadedBy: Optional[str] = None
    value: Optional[str] = None


class ObservationNewRequest(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    collector: Optional[str] = None
    geoCoordinates: Optional[GeoJSON] = None
    germplasmDbId: Optional[str] = None
    germplasmName: Optional[str] = None
    germplasmPUI: Optional[str] = None
    observationDbId: str
    observationTimeStamp: Optional[str] = None
    observationUnitDbId: Optional[str] = None
    observationUnitName: Optional[str] = None
    observationUnitPUI: Optional[str] = None
    observationVariableDbId: Optional[str] = None
    observationVariableName: Optional[str] = None
    observationVariablePUI: Optional[str] = None
    seasonDbId: Optional[str] = None
    seasonName: Optional[str] = None
    studyDbId: Optional[str] = None
    studyName: Optional[str] = None
    studyPUI: Optional[str] = None
    uploadedBy: Optional[str] = None
    value: Optional[str] = None


class PlannedCross(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    crossType: Optional[CrossType] = None
    crossingProjectDbId: Optional[str] = None
    crossingProjectName: Optional[str] = None
    parent1: Optional[CrossParent] = None
    parent2: Optional[CrossParent] = None
    plannedCrossDbId: str
    plannedCrossName: Optional[str] = None
    status: Optional[Status] = Field(
        None,
        description="The status of this planned cross. Is it waiting to be performed ('TODO'), has it been completed successfully ('DONE'), or has it not been done on purpose ('SKIPPED').",
    )


class Scale(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    dataType: Optional[DataType3] = Field(
        None,
        description='<p>Class of the scale, entries can be</p>\n<p>"Code" -  This scale class is exceptionally used to express complex traits. Code is a nominal scale that combines the expressions of the different traits composing the complex trait. For example a severity trait might be expressed by a 2 digit and 2 character code. The first 2 digits are the percentage of the plant covered by a fungus and the 2 characters refer to the delay in development, e.g. "75VD" means "75 %" of the plant is infected and the plant is very delayed.</p>\n<p>"Date" - The date class is for events expressed in a time format, See ISO 8601</p>\n<p>"Duration" - The Duration class is for time elapsed between two events expressed in a time format, e.g. days, hours, months</p>\n<p>"Nominal" - Categorical scale that can take one of a limited and fixed number of categories. There is no intrinsic ordering to the categories</p>\n<p>"Numerical" - Numerical scales express the trait with real numbers. The numerical scale defines the unit e.g. centimeter, ton per hectare, branches</p>\n<p>"Ordinal" - Ordinal scales are scales composed of ordered categories</p>\n<p>"Text" - A free text is used to express the trait.</p>',
    )
    decimalPlaces: Optional[int] = None
    ontologyReferenceDbId: Optional[str] = None
    scaleName: str
    scalePUI: Optional[str] = None
    units: Optional[str] = None
    validValues: Optional[ValidValues] = None


class ScaleNewRequest(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    dataType: Optional[DataType3] = Field(
        None,
        description='<p>Class of the scale, entries can be</p>\n<p>"Code" -  This scale class is exceptionally used to express complex traits. Code is a nominal scale that combines the expressions of the different traits composing the complex trait. For example a severity trait might be expressed by a 2 digit and 2 character code. The first 2 digits are the percentage of the plant covered by a fungus and the 2 characters refer to the delay in development, e.g. "75VD" means "75 %" of the plant is infected and the plant is very delayed.</p>\n<p>"Date" - The date class is for events expressed in a time format, See ISO 8601</p>\n<p>"Duration" - The Duration class is for time elapsed between two events expressed in a time format, e.g. days, hours, months</p>\n<p>"Nominal" - Categorical scale that can take one of a limited and fixed number of categories. There is no intrinsic ordering to the categories</p>\n<p>"Numerical" - Numerical scales express the trait with real numbers. The numerical scale defines the unit e.g. centimeter, ton per hectare, branches</p>\n<p>"Ordinal" - Ordinal scales are scales composed of ordered categories</p>\n<p>"Text" - A free text is used to express the trait.</p>',
    )
    decimalPlaces: Optional[int] = None
    ontologyReferenceDbId: Optional[str] = None
    scaleDbId: str
    scaleName: str
    scalePUI: Optional[str] = None
    units: Optional[str] = None
    validValues: Optional[ValidValues] = None


class Variable(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    commonCropName: Optional[str] = None
    contextOfUse: Optional[List[str]] = None
    defaultValue: Optional[str] = None
    documentationURL: Optional[str] = None
    growthStage: Optional[str] = None
    institution: Optional[str] = None
    language: Optional[str] = None
    method: Method
    ontologyReference: Optional[OntologyReference] = None
    scale: Scale
    scientist: Optional[str] = None
    status: Optional[str] = None
    submissionTimestamp: Optional[str] = None
    synonyms: Optional[List[str]] = None
    trait: Trait


class MetadataBase(BaseModel):
    datafiles: Optional[List[DataFile]] = Field(
        None,
        description='The datafiles contains a list of file URLs and metadata. \nThese files contain additional information related to the returned object and can be retrieved by a subsequent call. \nThis could be a supplementary data file, an informational file, the uploaded file where the data originated from, a generated file representing the whole dataset in a particular format, or any other related file. ',
        examples=[[]],
    )
    status: Optional[List[StatusModel]] = Field(
        None,
        description='The status field contains a list of informational status messages from the server. \nIf no status is reported, an empty list should be returned. See Error Reporting for more information.',
    )


class MetadataTokenPagination(MetadataBase):
    pagination: Optional[TokenPagination] = None


class Attribute(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    attributeCategory: Optional[str] = None
    attributeDbId: Optional[str] = None
    attributeDescription: Optional[str] = None
    attributeName: str
    attributePUI: Optional[str] = None
    commonCropName: Optional[str] = None
    contextOfUse: Optional[List[str]] = None
    defaultValue: Optional[str] = None
    documentationURL: Optional[str] = None
    growthStage: Optional[str] = None
    institution: Optional[str] = None
    language: Optional[str] = None
    method: Method
    ontologyReference: Optional[OntologyReference] = None
    scale: Scale
    scientist: Optional[str] = None
    status: Optional[str] = None
    submissionTimestamp: Optional[str] = None
    synonyms: Optional[List[str]] = None
    trait: Trait


class Cross(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    crossName: Optional[str] = None
    crossType: Optional[CrossType] = None
    crossingProjectDbId: Optional[str] = None
    crossingProjectName: Optional[str] = None
    parent1: Optional[CrossParent] = None
    parent2: Optional[CrossParent] = None
    plannedCrossDbId: Optional[str] = None
    plannedCrossName: Optional[str] = None


class CrossNewRequest(BaseModel):
    additionalInfo: Optional[AdditionalInfo] = None
    crossDbId: str
    crossName: Optional[str] = None
    crossType: Optional[CrossType] = None
    crossingProjectDbId: Optional[str] = None
    crossingProjectName: Optional[str] = None
    parent1: Optional[CrossParent] = None
    parent2: Optional[CrossParent] = None
    plannedCrossDbId: Optional[str] = None
    plannedCrossName: Optional[str] = None


class Metadata(MetadataBase):
    pagination: Optional[BasePagination] = None
